program/rust/src/rust_oracle.rs

```diff
@@ -32,47 +40,154 @@ use crate::c_oracle_header::{
+        let mut price_account = load_account_as_mut::<PriceAccountWrapper>(price_account_info)?;
+        price_account.add_price_to_time_machine()?;
+        msg!("updated tracker!");
+    }
+    Ok(c_ret_value)
+}
+fn send_lamports<'a>(
+    from: &AccountInfo<'a>,
+    to: &AccountInfo<'a>,
+    system_program: &AccountInfo<'a>,
+    ammount: u64,
+) -> Result<(), ProgramError> {
+    let transfer_instruction = transfer(from.key, to.key, ammount);
+    invoke(
+        &transfer_instruction,
+        &[from.clone(), to.clone(), system_program.clone()],
+    )?;
+    Ok(())
+}
+
+///resize price account and initialize the TimeMachineStructure
+pub fn upgrade_price_account<'a>(
+    funding_account_info: &AccountInfo<'a>,
+    price_account_info: &AccountInfo<'a>,
+    system_program: &AccountInfo<'a>,
+    program_id: &Pubkey,
+) -> OracleResult {
+    check_valid_funding_account(funding_account_info)?;
+    check_valid_signable_account(program_id, price_account_info, size_of::<pc_price_t>())?;
+    let account_len = price_account_info.try_data_len()?;
+    match account_len {
+        PRICE_T_SIZE => {
+            //ensure account is still rent exempt after resizing
+            let rent: Rent = Default::default();
+            let lamports_needed: u64 = rent
+                .minimum_balance(size_of::<PriceAccountWrapper>())
+                .saturating_sub(price_account_info.lamports());
+            if lamports_needed > 0 {
+                send_lamports(
+                    funding_account_info,
+                    price_account_info,
+                    system_program,
+                    lamports_needed,
+                )?;
+            }
+            //resize
+            //we do not need to zero initialize since this is the first time this memory
+            //is allocated
+            price_account_info.realloc(size_of::<PriceAccountWrapper>(), false)?;
+            //update twap_tracker
+            let mut price_account = load_account_as_mut::<PriceAccountWrapper>(price_account_info)?;
+            price_account.invalidate_current_time_machine_entries()?;
+            Ok(SUCCESS)
+        }
+        PRICE_ACCOUNT_SIZE => Ok(SUCCESS),
+        _ => Err(ProgramError::InvalidArgument),
+    }
+}
+
+fn get_account_type(account: &AccountInfo) -> Result<u32, ProgramError> {
+    let account_type = load_account_as::<pc_acc>(account)?.type_;
+    Ok(account_type)
 }
-/// has version number/ account type dependant logic to make sure the given account is compatible
-/// with the current version
+
+/// We should call this instruction on any account that needs to be
+/// upgraded after an Oracle update
 /// updates the version number for all accounts, and resizes price accounts
-pub fn update_version(
-    _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
+/// accounts[0] funding account           [signer writable]
+/// accounts[1] upgradded acount       [signer writable]
+/// accounts [2] system program
+pub fn upgrade_account(
```

Comment:
fwiw i'm not sure it makes sense to have a generic `upgrade_account` instruction. We don't do account upgrades often, and this isn't saving us much code even if we add another account upgrade in the future.

Given that you already wrote it, this is fine, but figured i'd share my thoughts.

Reply:
Yep, it's going to become upgrade_price_account